File: src/common/543/687.py
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def dfs(node: Optional[TreeNode]) -> int:
            if node is None:
                return -1  # 下面 +1 后，对于叶子节点就刚好是 0
            l_len = dfs(node.left) + 1  # 左子树最大链长+1
            r_len = dfs(node.right) + 1  # 右子树最大链长+1
            if node.left and node.left.val != node.val: l_len = 0  # 视作 0
            if node.right and node.right.val != node.val: r_len = 0  # 视作 0
            nonlocal ans
            ans = max(ans, l_len + r_len)  # 两条链拼成路径
            return max(l_len, r_len)  # 当前子树最大链长
        dfs(root)
        return ans

File: src/draft/0a0draft1201/687.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestUnivaluePath = function (root) {
  let ans = 0;
  const dfs = (root) => {
    if (root == null) return 0;
    let leftDepth = dfs(root.left);
    let rightDepth = dfs(root.right);
    if (root.left && root.left.val != root.val) leftDepth = 0;
    if (root.right && root.right.val != root.val) rightDepth = 0;
    ans = Math.max(ans, leftDepth + rightDepth);
    return Math.max(leftDepth, rightDepth) + 1;
  };
  dfs(root);
  return ans;
};
